pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go

```diff
@@ -0,0 +1,247 @@
+		}
+
+		return nil, err
+	}
+
+	return &CouchDBEDVProvider{coreProvider: couchDBProvider}, nil
+}
+
+// CreateStore creates a new store with the given name.
+func (c CouchDBEDVProvider) CreateStore(name string) error {
+	return c.coreProvider.CreateStore(name)
+}
+
+// OpenStore opens an existing store and returns it.
+func (c CouchDBEDVProvider) OpenStore(name string) (edvprovider.EDVStore, error) {
+	coreStore, err := c.coreProvider.OpenStore(name)
+	if err != nil {
+		return nil, err
+	}
+
+	return &CouchDBEDVStore{coreStore: coreStore}, nil
+}
+
+// CouchDBEDVStore represents a CouchDB store with functionality needed for EDV data storage.
+// It wraps an edge-core CouchDB store with additional functionality that's needed for EDV operations.
+type CouchDBEDVStore struct {
+	coreStore storage.Store
+}
+
+// Put stores the given document.
+// A mapping document is also created and stored in order to allow for encrypted indices to work.
+func (c CouchDBEDVStore) Put(document models.EncryptedDocument) error {
+	documentBytes, err := json.Marshal(document)
+	if err != nil {
+		return err
+	}
+
+	// TODO: The encrypted document and mapping document should both be stored at the same time (all-or-nothing).
+	// If either of these requests fails, then the database will be left in a weird state.
+	// https://github.com/trustbloc/edge-core/issues/27 and https://github.com/trustbloc/edv/issues/49
+
+	for _, indexedAttributeCollection := range document.IndexedAttributeCollections {
+		for _, indexedAttribute := range indexedAttributeCollection.IndexedAttributes {
+			err := c.createMappingDocument(indexedAttribute.Name, document.ID)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return c.coreStore.Put(document.ID, documentBytes)
+}
+
+// createMappingDocument creates a document with a mapping of the encrypted index to the document that has it.abcdefgh
+func (c CouchDBEDVStore) createMappingDocument(indexedAttributeName, encryptedDocID string) error {
+	mapDocument := couchDBIndexMappingDocument{
+		IndexName:              indexedAttributeName,
+		MatchingEncryptedDocID: encryptedDocID,
+	}
+
+	documentBytes, err := json.Marshal(mapDocument)
+	if err != nil {
+		return err
+	}
+
+	err = c.coreStore.Put(encryptedDocID+"_mapping_"+uuid.New().String(), documentBytes)
+	if err != nil {
+		return err
+	}
+
+	createIndexRequest := storage.CreateIndexRequest{
+		IndexStorageLocation: "EDV_EncryptedIndexesDesignDoc",
+		IndexName:            "EDV_IndexName",
+		WhatToIndex:          `{"fields": ["` + mapDocumentIndexedField + `"]}`,
+	}
+
+	return c.coreStore.CreateIndex(createIndexRequest)
+}
+
+// Get fetches the document associated with the given key.
+func (c CouchDBEDVStore) Get(k string) ([]byte, error) {
+	return c.coreStore.Get(k)
+}
+
+// CreateEDVIndex creates the index which will allow for encrypted indices to work.
+func (c CouchDBEDVStore) CreateEDVIndex() error {
+	createIndexRequest := storage.CreateIndexRequest{
```

Comment:
This code is a copy of the code in createMappingDocument.

Reply:
Fixed - see https://github.com/trustbloc/edv/pull/51#discussion_r405010881